pywbem_mock/_mainprovider.py

```diff
@@ -0,0 +1,4209 @@
+
+        if result_class:
+            self._validate_class_exists(namespace, result_class, "ResultClass")
+
+        result_classes = self._subclasses_lc(result_class, class_store)
+        assoc_classes = self._subclasses_lc(assoc_class, class_store)
+
+        inst_name.namespace = namespace
+        role = role.lower() if role else role
+        result_role = result_role.lower() if result_role else result_role
+
+        ref_paths = self._get_reference_instnames(namespace, inst_name,
+                                                  assoc_class, role)
+
+        # Get associated instance names
+        rtn_instpaths = set()
+        for ref_path in ref_paths:
+            inst = self._find_instance(ref_path, instance_store)
+            for prop in six.itervalues(inst.properties):
+                if prop.type == 'reference':
+                    if prop.value == inst_name:
+                        if assoc_class \
+                                and inst.classname.lower() not in assoc_classes:
+                            continue
+                        if role and prop.name.lower() != role:
+                            continue
+                    else:
+                        if result_class and (prop.value.classname.lower()
+                                             not in result_classes):
+                            continue
+                        if result_role and prop.name.lower() != result_role:
+                            continue
+                        rtn_instpaths.add(prop.value)
+        return rtn_instpaths
+
+    ####################################################################
+    #
+    #   ReferenceNames server request
+    #
+    ####################################################################
+
+    def ReferenceNames(self, namespace, ObjectName, ResultClass=None,
+                       Role=None):
+        """
+        Implements a WBEM server responder for
+        :meth:`~pywbem.WBEMConnection.ReferenceNames`.
+
+        Retrieves the instance paths of the association instances that
+        reference a source instance, or the class paths of the association
+        classes that reference a source class from the CIM instances in the CIM
+        repository. The association instances defining the relationship between
+        ObjectName and result classes must exist in the CIM repository.
+
+        Parameters:
+
+          namespace (:term:`string`):
+            The name of the CIM namespace in the CIM repository (case
+            insensitive). Must not be `None`. Leading or trailing slash
+            characters are ignored.
+
+          ObjectName:
+            The object path of the source object:
+
+            * Instance-level use: The instance path of the
+              source instance, as a :class:`~pywbem.CIMInstanceName` object.
+              If this object does not specify a namespace, the default namespace
+              of the connection is used.
```

Comment:
On "the default namespace of the connection": Change to "the namespace specified in the namespace parameter".

Reply:
DONE